zipkin-sampler/src/main/scala/com/twitter/zipkin/sampler/AdaptiveSampleRate.scala

```diff
@@ -0,0 +1,373 @@
+/*
+ * Copyright 2012 Twitter Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+package com.twitter.zipkin.sampler
+
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger, AtomicLong, AtomicReference}
+
+import com.google.common.collect.EvictingQueue
+import com.twitter.conversions.time._
+import com.twitter.finagle.stats.{DefaultStatsReceiver, StatsReceiver}
+import com.twitter.finagle.util.DefaultTimer
+import com.twitter.logging.Logger
+import com.twitter.util._
+
+import scala.reflect.ClassTag
+
+/**
+ * Dynamically adjusts [[boundary]] derived from a target rate.
+ *
+ * This node is responsible for incrementing [[spanCount]] whenever a span is
+ * sent to storage. This function will watch [[spanCount]] value and feed it
+ * back into a global storage rate. A result might be a lower sample rate,
+ * which means a lower trace ID [[boundary]].
+ */
+class AdaptiveSampleRate(
```

Comment:
This class was extracted from AdaptiveSampler. Notably, it coordinates with Atomic* and addresses the observer stuff (Var etc) internally. This allows it to be easily used with java.